src/librustc_lint/builtin.rs

```diff
@@ -2116,6 +2116,87 @@ impl ClashingExternDecl {
             SymbolName::Normal(fi.ident.name)
         }
     }
+
+    /// Checks whether two types are structurally the same enough that the declarations shouldn't
+    /// clash. We need this so we don't emit a lint when two modules both declare an extern struct,
+    /// with the same members (as the declarations shouldn't clash).
+    fn structurally_same_type<'tcx>(tcx: TyCtxt<'tcx>, a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
```

Comment:
Hmm, I've just realised that this isn't adequate because it doesn't take layout modifiers into account (eg, if a type is packed, or a different alignment is specified for one of the structs). Going to take a look into comparing layout information instead.